cpp/src/gandiva/cache.h

```diff
@@ -31,29 +34,50 @@ int GetCapacity();
 GANDIVA_EXPORT
 void LogCacheSize(size_t capacity);
 
+GANDIVA_EXPORT
+int GetCacheTypeToUse();
+
 template <class KeyType, typename ValueType>
 class Cache {
  public:
-  explicit Cache(size_t capacity) : cache_(capacity) { LogCacheSize(capacity); }
+  explicit Cache(size_t capacity, int cache_type_to_use) {
+    if (cache_type_to_use == 0) {
+      this->cache_ =
+          std::make_unique<GreedyDualSizeCache<KeyType, ValueObject>>(capacity);
+    } else {
+      this->cache_ = std::make_unique<LruCache<KeyType, ValueObject>>(capacity);
+    }
+    LogCacheSize(capacity);
+  }
 
-  Cache() : Cache(GetCapacity()) {}
+  Cache() : Cache(GetCapacity(), GetCacheTypeToUse()) {}
 
   ValueType GetModule(KeyType cache_key) {
-    arrow::util::optional<ValueType> result;
+    arrow::util::optional<ValueObject> result;
     mtx_.lock();
-    result = cache_.get(cache_key);
+    result = (*cache_).get(cache_key);
     mtx_.unlock();
-    return result != arrow::util::nullopt ? *result : nullptr;
+    return result != arrow::util::nullopt ? (*result).module : nullptr;
   }
 
-  void PutModule(KeyType cache_key, ValueType module) {
+  void PutModule(KeyType cache_key, ValueType module, uint64_t priority) {
```

Comment:
@projjal 
I believe our cache implementations should be the most generic possible. I do not think it would be good to enforce us to always create a "Struct Object" in the Projector::Make and the Filter::Make just to use the cache. 
This is the reason why I prefer to create this struct "ValueObject" on this cache.h abstraction and not on the Projector.cc and Filter.cc while using the cache.

What do you think? If you think it is better to enforce the "Struct Object" usage on the Projector and Filter Make methods, I can do it.

Reply:
How is this generic? If we have a float type cost in the future? Or more than one properties. You have to change all the apis again.

Reply:
Ok, so I will change all our cache implementations to always receive a Value of a specific type "Struct Object", correct?
I will work on it right now